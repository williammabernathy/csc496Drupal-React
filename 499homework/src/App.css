/*
  Custom
*/
.error
{
  max-width: 500px;
  margin: auto;
  font-size: 50px;
  padding-top: 40px;
}

.sepCol{
  display: flex; 
  flex-wrap: wrap;
  margin: 15px 15px 15px 15px;
}

.card {
  border: round;
  border-style: solid;
  margin: 0 20px 15px 20px;
}

.App {
  background-color: #e6eeff;
}

.content {
  background-color: white;
  margin: 0 100px 0 100px;
}

h1 {
  font-family: 'Trocchi', serif;
  font-weight: normal;
  padding-top: 10px;
  padding-left: 15px;
}

.footer {
  background-color: #f2f2f2;
  border-radius: 3px;
  padding-top: 10px;
  padding-right: 10px;
  padding-bottom: 10px;
  padding-left: 10px;
  margin: 10px 0 0 0;
}

.footerTable {
  margin: 8px 8px 8px 8px;
}

/*
  Full Recipe Details
*/
.previous {
  float: left;
  margin: 10px 0 0 0;
  background-color: #f2f2f2;
}

.next {
  float: right;
  margin: 10px 0 0 0;
  background-color: #f2f2f2;
}

.cardFull {
  border-style: solid;
  border-width: 1px;
  border-color: lightgrey;
  border-radius: 10px;
  padding-top: 10px;
  padding-right: 10px;
  padding-bottom: 10px;
  padding-left: 10px;
  margin: 10px 10px 10px 10px;
  overflow: hidden;
}

.cardFullImage {
  height:auto;
}

/*
  NavBar
*/
.mainMenu {
  border-style: solid;
  border-width: 1px;
  border-color: lightgrey;
  border-radius: 3px;
}

.navLink {
  font-size: 20px;
  margin: 10px 10px 10px 10px;
}

/*
  Generated

.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #09d3ac;
}
*/
