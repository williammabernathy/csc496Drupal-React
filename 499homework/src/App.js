import React, { Component } from 'react';
import { Navbar, Nav } from 'react-bootstrap';
import { Card, CardImg, CardText, CardBody, CardTitle, Button } from 'reactstrap';
import './App.css';
import 'bootstrap/dist/css/bootstrap.min.css';

/*
  Variables/Constants
*/
const fetch = require("node-fetch");

//api url variables
const PATH_BASE = 'http://gtest.dev.wwbtc.com'; 
const PATH_JSON = '/json'; 
const PARAM_TYPE = '/rec';

//variables for changing url
const recURL = "#recipe";
const homeURL = "#home";

/*
  Components
*/
//display a basic card of individual recipes
const Recipe = ( { ID, title, field_images, field_summary, viewRecipeClick } ) =>
{
  //split images, only want the first of two
  const image = field_images.split(',').slice(0)[0].trim();

  return (
    <div className="recipeStyle">
      <Card className = "card" style={{ width: '18rem', border: "1px solid lightgrey" }}>
        <CardImg top width="25%" src={`${PATH_BASE}${image}`} alt="Food Image" />
        <CardBody>
          <CardTitle><b>{title}</b></CardTitle>
          <CardText><div dangerouslySetInnerHTML={{ __html: field_summary }}/></CardText>

          {/* whenever an element is clicked, get its ID */}
          <center>
            <Button href={`${recURL}/${ID}`} onClick={() => viewRecipeClick(ID)}>
              View Recipe
            </Button>
          </center>

        </CardBody>
      </Card>
    </div>
  );
}

//display a recipe's full details
const RecipeFull = ({ ID, title, field_images, field_ingredients, body, changeRecipe }) =>
{
  //split images, only want the first of two
  const image = field_images.split(',').slice(0)[0].trim();

  return(
    <div className="cardFull" >
      <div className="cardFullImage"><img src={`${PATH_BASE}${image}`} alt="Food" /></div>
      <div>
      <hr/>
        <div><b>{title}</b></div><br/>
        <div>{field_ingredients}</div><br/>
        <div><div dangerouslySetInnerHTML={{ __html: body }}/></div>
        <hr/>

        {/* buttons to cycle through recipes. change url when pressed*/}
        <Button href={`${recURL}/${ID}`} className="previous" onClick = {() => changeRecipe(ID-1)}>
          Previous
        </Button>
        <Button href={`${recURL}/${ID}`} className="next" onClick = {() => changeRecipe(ID+1)}>
          Next
        </Button>
      </div>
    </div>
  );
}

const HomeHeading = ( { body } ) =>
{
  return (
    <div>
      <div dangerouslySetInnerHTML={{ __html: body }}/>
    </div>
  );
}

//loading screen
//currently only used when page is first loaded
const Loading = () =>
{
  return (
    <div className="error">
      Loading..
    </div>
  );
}

//our header containing the nav bar
const Header = ({ viewHome, viewRecipeNavClick }) => 
{
  return (
    <Navbar bg="light" expand="lg" className="mainMenu">
      <Navbar.Brand className="navBrand" href={homeURL} onClick={() => viewHome()}>CSC496</Navbar.Brand>
      <Nav>
        <Nav.Link className="navLink" href={homeURL} onClick={() => viewHome()}>Home</Nav.Link>
        <Nav.Link className="navLink" href={recURL} onClick={() => viewRecipeNavClick()}>Recipes</Nav.Link>
        <Nav.Link className="navLink" href="https://github.com/williammabernathy/csc496Drupal-React">Github</Nav.Link>
      </Nav>
    </Navbar>
  );
}

//general footer
const Footer = () =>
{
  return(
    <div className = "footer">
      <span>
        CSC 496 Project
      </span> <br/>
      <span>
        by William Abernathy
      </span>
    </div>
  );
}

/*
  Main
*/
class App extends Component 
{
  constructor(props) 
  {
    super(props);

    this.state = {
      results: null,            //data from fetching recipies
      page: "home",             //what page we're currently on
      currentRecipe: null,      //which recipe has been selected; the recipe ID
      homeHead: null,           //data from fetching home page
    }

    this.fetchRecipes = this.fetchRecipes.bind(this);
    this.fetchHome = this.fetchHome.bind(this);
    this.viewRecipeClick = this.viewRecipeClick.bind(this);
    this.viewHome = this.viewHome.bind(this);
    this.viewRecipeNavClick = this.viewRecipeNavClick.bind(this);
    this.changeRecipe = this.changeRecipe.bind(this);
  }

  /*
    Functions
  */
 //api call
  fetchRecipes()
  {
    fetch(`${PATH_BASE}${PATH_JSON}${PARAM_TYPE}`)
    .then(response => response.json())  
    .then(results => this.setState({ results }))
  }

  fetchHome()
  {
    fetch(`http://gtest.dev.wwbtc.com/json/page?_format=json`)
    .then(response => response.json())
    .then(homeHead => this.setState({ homeHead }))
  }

  componentDidMount() 
  { 
    this.fetchRecipes(); 
    this.fetchHome();
  }

  //the view more details on a recipe card was pressed, show that recipe's
  //full detailed page
  viewRecipeClick(index)
  {
    this.setState({ page: "recipeDetailed", currentRecipe: index })
  }

  //nav bar home button was pressed, show home page
  viewHome()
  {
    this.setState({ page: "home", currentRecipe: null });
  }

  //nav bar recipe button was pressed, show full list of recipes
  viewRecipeNavClick()
  {
    this.setState({ page: "recipes", currentRecipe: null });
  }

  //update the currentRecipe we're currently viewing once previous or next is pressed
  changeRecipe(ID)
  {
    const { results } = this.state;

    //if -1, means previous was pressed. start over from last element in results
    if(ID === -1)
    {
      this.setState({ currentRecipe: results.length - 1 });
    }
    //if we're at max and the length is reached, start over from beginning
    else if (ID === results.length)
    {
      this.setState({ currentRecipe: 0 })
    }
    else
    {
      this.setState({ currentRecipe: ID })
    }
  }

  /*
    Rendered Elements
  */
  render() 
  {
    const { results, page, currentRecipe, homeHead} = this.state;

    //if results null, show loading
    //first pass always returns null
    if (!results || !homeHead) 
    {
      return (
        <div className="App">
          <div className="content">
            <Header
              viewHome={this.viewHome}
              viewRecipeNavClick={this.viewRecipeNavClick}
            />

            <Loading />

            <Footer />
          </div>
        </div>
      );
    }
    else 
    {
      //show home page
      if (page === "home") 
      {
        return (
          <div className="App">
            <div className="content">
              <Header
                viewHome={this.viewHome}
                viewRecipeNavClick={this.viewRecipeNavClick}
              />
              <div className="homeHeader">
                {homeHead.map((homeHeading, index) => (
                  <HomeHeading 
                    key={index}
                    body={homeHeading["body"]} 
                  />
                ))}
              </div>

              <hr />

              <h1>Recipes</h1>

              <div className="sepCol">
                {results.map((recipes, index) => (
                  <Recipe
                    key={index}
                    ID={index}
                    title={recipes["title"]}
                    field_images={recipes["field_images"]}
                    field_summary={recipes["field_summary"]}
                    viewRecipeClick={this.viewRecipeClick}
                  />
                ))}
              </div>

              <hr />

              <h1>Articles</h1>
              ToDo
              <br />
              <br />
              <br />

              <Footer />
            </div>
          </div>
        );
      }
      //show recipes page
      else if (page === "recipes") 
      {
        return (
          <div className="App">
            <div className="content">
              <Header
                viewHome={this.viewHome}
                viewRecipeNavClick={this.viewRecipeNavClick}
              />

              <h1>All Our Recipes</h1>

              <div className='sepCol'>
                {results.map((recipes, index) =>
                  <Recipe key={index}
                    ID={index}
                    title={recipes['title']}
                    field_images={recipes['field_images']}
                    field_summary={recipes['field_summary']}
                    viewRecipeClick={this.viewRecipeClick}
                  />
                )}
              </div>

              <Footer />
            </div>
          </div>
        );
      }
      //show detailed individual recipe page
      else if (page === "recipeDetailed") 
      {
        return (
          <div className="App">
            <div className="content">
              <Header
                viewHome={this.viewHome}
                viewRecipeNavClick={this.viewRecipeNavClick}
              />

              {results.slice(currentRecipe, currentRecipe + 1).map((recipes) =>
                <RecipeFull 
                  ID={currentRecipe}
                  title={recipes['title']}
                  field_images={recipes['field_images']}
                  field_ingredients={recipes['field_ingredients']}
                  body={recipes['body']}
                  changeRecipe={this.changeRecipe}
                />
              )}

              <Footer />
            </div>
          </div>
        );
      }
    }
  }
}

export default App;