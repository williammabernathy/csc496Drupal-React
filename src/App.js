import React, { Component } from 'react';
import { Navbar, Nav } from 'react-bootstrap';
import { Card, CardImg, CardText, CardBody, CardTitle, Button } from 'reactstrap';
import './App.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import { fetch, PATH_BASE, PATH_JSON, PARAM_TYPE, recURL, homeURL } from './constants/index.js'

//display a basic card of individual recipes
const Recipe = ( { ID, title, field_images, field_summary, viewRecipeClick } ) =>
{
  //split images, only want the first of two
  if(field_images)
  {
    var image = field_images.split(',').slice(0)[0].trim();
  }

  return (
    <div className="recipeStyle">
      <Card className = "card" style={{ width: '18rem', border: "1px solid lightgrey" }}>
        <CardImg top width="25%" src={`${PATH_BASE}${image}`} alt="Food Image" />
        <CardBody>
          <CardTitle><b>{title}</b></CardTitle>
          <CardText><div dangerouslySetInnerHTML={{ __html: field_summary }}/></CardText>

          {/* whenever an element is clicked, get its ID */}
          <center>
            <Button id="full" href={`${recURL}/${ID}`} onClick={() => viewRecipeClick(ID)}>
              View Recipe
            </Button>
          </center>

        </CardBody>
      </Card>
    </div>
  );
}

//display a recipe's full details
const RecipeFull = ({ ID, title, field_images, field_ingredients, body, changeRecipe }) =>
{
  //split images, only want the first of two
  if(field_images)
  {
    var image = field_images.split(',').slice(0)[0].trim();
  }

  return(
    <div className="cardFull" >
      <div className="cardFullImage"><img src={`${PATH_BASE}${image}`} alt="Food" /></div>
      <div>
      <hr/>
        <div><b>{title}</b></div><br/>
        <div>{field_ingredients}</div><br/>
        <div><div dangerouslySetInnerHTML={{ __html: body }}/></div>
        <hr/>

        {/* buttons to cycle through recipes. change url when pressed
            increment or decrement ID to cycle recipes */}
        <Button id="previous" href={`${recURL}/${ID}`} className="previous" onClick = {() => changeRecipe(ID-1)}>
          Previous
        </Button>
        <Button id="next" href={`${recURL}/${ID}`} className="next" onClick = {() => changeRecipe(ID+1)}>
          Next
        </Button>
      </div>
    </div>
  );
}

//Head component used to display data from fetching the home page
//info from the test drupal site
//renders the html within the json data
const HomeHeading = ( { body } ) =>
{
  return (
    <div>
      <div dangerouslySetInnerHTML={{ __html: body }}/>
    </div>
  );
}

//loading screen
//currently only used when page is first loaded
const Loading = () =>
{
  return (
    <div className="error">
      Loading..
    </div>
  );
}

//our header containing the nav bar
const Header = ({ viewHome, viewRecipeNavClick }) => 
{
  return (
    <Navbar bg="light" expand="lg" className="mainMenu">
      <Navbar.Brand id="brand" className="navBrand" href={homeURL} onClick={() => viewHome()}>CSC496</Navbar.Brand>
      <Nav>
        <Nav.Link id="viewHome" className="navLink" href={homeURL} onClick={() => viewHome()}>Home</Nav.Link>
        <Nav.Link id="viewRecList" className="navLink" href={recURL} onClick={() => viewRecipeNavClick()}>Recipes</Nav.Link>
        <Nav.Link id="viewGit" className="navLink" href="https://github.com/williammabernathy/csc496Drupal-React">Github</Nav.Link>
      </Nav>
    </Navbar>
  );
}

//general footer component
const Footer = () =>
{
  return(
    <div className = "footer">
      <span>
        CSC 496 Project
      </span> <br/>
      <span>
        by William Abernathy
      </span>
    </div>
  );
}

/*
  Main
*/
class App extends Component 
{
  constructor(props) 
  {
    super(props);

    this.state = {
      results: null,            //data from fetching recipies
      page: "home",             //what page we're currently on
      currentRecipe: null,      //which recipe has been selected; the recipe ID
      homeHead: null,           //data from fetching home page
    }

    //binding functions
    this.fetchRecipes = this.fetchRecipes.bind(this);
    this.fetchHome = this.fetchHome.bind(this);
    this.viewRecipeClick = this.viewRecipeClick.bind(this);
    this.viewHome = this.viewHome.bind(this);
    this.viewRecipeNavClick = this.viewRecipeNavClick.bind(this);
    this.changeRecipe = this.changeRecipe.bind(this);
  }

  /*
    Functions
  */
 //fetch recipes from test drupal site
  fetchRecipes()
  {
    fetch(`${PATH_BASE}${PATH_JSON}${PARAM_TYPE}`).then(response => response.json()).then(results => this.setState({ results }))
  }

  //fetch home page information from drupal test site
  fetchHome()
  {
    fetch(`http://gtest.dev.wwbtc.com/json/page?_format=json`).then(response => response.json()).then(homeHead => this.setState({ homeHead }))
  }

  //once component mounts on second pass, fetch data
  componentDidMount() 
  { 
    this.fetchRecipes(); 
    this.fetchHome();
  }

  //the view more details on a recipe card was pressed, show that recipe's
  //full detailed page
  viewRecipeClick(index)
  {
    this.setState({ page: "recipeDetailed", currentRecipe: index })
  }

  //nav bar home button was pressed, show home page
  viewHome()
  {
    this.setState({ page: "home", currentRecipe: null });
  }

  //nav bar recipe button was pressed, show full list of recipes
  viewRecipeNavClick()
  {
    this.setState({ page: "recipes", currentRecipe: null });
  }

  //update the currentRecipe we're currently viewing once previous or next is pressed
  //return statements added for testing purposes
  changeRecipe(ID)
  {
    const { results, currentRecipe } = this.state;

    //if -1, means previous was pressed while at first index. 
    //start over from last element in results
    if(results && ID === -1)
    {
      this.setState({ currentRecipe: results.length - 1 });
      return this.state.currentRecipe;
    }
    //if we're at max and the length is reached, start over from beginning
    else if (results && ID === results.length)
    {
      this.setState({ currentRecipe: 0 })
      return this.state.currentRecipe;
    }
    else
    {
      this.setState({ currentRecipe: ID })
      return this.state.currentRecipe;
    }
  }

  /*
    Rendered Elements
  */
  render() 
  {
    //get our data from props to use for rendering
    const { results, page, currentRecipe, homeHead} = this.state;

    //slice up the results and only take first 2 to display on the homepage
    if(results)
    {
      var resultsRecHome = results.slice(0, 2);
    }

    //only want first homepage message
    if(homeHead)
    {
      var homeHeadSlice = homeHead.slice(0, 1);
    }

    //if results null, show loading
    //first pass always returns null as App component not mounted
    if (!results || !homeHead) 
    {
      return (
        <div className="App">
          <div className="content">
            <Header
              viewHome={this.viewHome}
              viewRecipeNavClick={this.viewRecipeNavClick}
            />

            <Loading />

            <Footer />
          </div>
        </div>
      );
    }
    else 
    {
      //show home page; default 
      if (page === "home") 
      {
        return (
          <div className="App">
            <div className="content">
              <Header
                viewHome={this.viewHome}
                viewRecipeNavClick={this.viewRecipeNavClick}
              />
              <div className="homeHeader">
                {homeHeadSlice.map((homeHeading, index) => (
                  <HomeHeading 
                    key={index}
                    body={homeHeading["body"]} 
                  />
                ))}
              </div>

              <hr />

              <h1>Recipes</h1>

              <div className="sepCol">
                {resultsRecHome.map((recipes, index) => (
                  <Recipe
                    key={index}
                    ID={index}
                    title={recipes["title"]}
                    field_images={recipes["field_images"]}
                    field_summary={recipes["field_summary"]}
                    viewRecipeClick={this.viewRecipeClick}
                  />
                ))}
              </div>

              <hr />

              <h1>Articles</h1>
              ToDo
              <br />
              <br />
              <br />

              <Footer />
            </div>
          </div>
        );
      }
      //show recipes page
      //page change occurs when navbar viewRecipe button is clicked
      else if (page === "recipes") 
      {
        return (
          <div className="App">
            <div className="content">
              <Header
                viewHome={this.viewHome}
                viewRecipeNavClick={this.viewRecipeNavClick}
              />

              <h1>All Our Recipes</h1>

              <div className='sepCol'>
                {results.map((recipes, index) =>
                  <Recipe key={index}
                    ID={index}
                    title={recipes['title']}
                    field_images={recipes['field_images']}
                    field_summary={recipes['field_summary']}
                    viewRecipeClick={this.viewRecipeClick}
                  />
                )}
              </div>

              <Footer />
            </div>
          </div>
        );
      }
      //show detailed individual recipe page
      //page change occurs when viewRecipeDetail is clicked
      //from homepage recipes or recipe lists
      else if (page === "recipeDetailed") 
      {
        return (
          <div className="App">
            <div className="content">
              <Header
                viewHome={this.viewHome}
                viewRecipeNavClick={this.viewRecipeNavClick}
              />

              {results.slice(currentRecipe, currentRecipe + 1).map((recipes) =>
                <RecipeFull 
                  ID={currentRecipe}
                  title={recipes['title']}
                  field_images={recipes['field_images']}
                  field_ingredients={recipes['field_ingredients']}
                  body={recipes['body']}
                  changeRecipe={this.changeRecipe}
                />
              )}

              <Footer />
            </div>
          </div>
        );
      }
    }
  }
}

export default App;

//export custom components for reference elsewhere
//ie testing
export { Recipe, RecipeFull, HomeHeading, Footer, Header, Loading, };