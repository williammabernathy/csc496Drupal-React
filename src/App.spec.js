import React from 'react';
import renderer from 'react-test-renderer';
import App, { Recipe, RecipeFull, HomeHeading, Footer, Header, Loading, } from './App';
import { Navbar, Nav} from 'react-bootstrap';
import { mount, shallow } from 'enzyme';

/* App Component (Main) */
describe('App', () => {
    const wrapper = shallow(<App />);
    var wrapperMount;

    //mock data to mimic results in App component
    var mockRecipe = [
        {
            body: "body",
            field_images: "image",
            field_ingredients: "1 111",
            field_summary: "summary",
            title: "title",
        },
        {
            body: "body",
            field_images: "image",
            field_ingredients: "1 1 1 1",
            field_summary: "summary",
            title: "title ",
        }
    ]

    //mock data to mimic home page results
    var mockHome = [
        {
            body: "body",
            title: "title",
        }
    ]

    //simple test to check if the component renders via snapshot
    test('App snapshot renders', () => 
    {
        const component = renderer.create(<App />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    //before each tests, mount the app component
    //not all tests will use mount, some will use shallow
    beforeEach(() => 
    {
        wrapperMount = mount(<App />);
    });

    //check that components inside App render
    //also check multiple pages
    it('renders Recipe on home page', () =>
    {
        wrapperMount.setState({ results: mockRecipe, homeHead: mockHome, page: "home" });
        //wrapper.instance().forceUpdate()
        expect(wrapperMount.find(Recipe).length).toEqual(2);
    });

    it('renders Recipe on recipe page', () =>
    {
        wrapperMount.setState({ results: mockRecipe, homeHead: mockHome, page: "recipes" });
        //wrapper.instance().forceUpdate()
        expect(wrapperMount.find(Recipe).length).toEqual(2);
    });

    it('renders RecipeFull current at 1', () =>
    {
        wrapperMount.setState({ results: mockRecipe, currentRecipe: 1, homeHead: mockHome, page: "recipeDetailed" });
        //wrapper.instance().forceUpdate()
        expect(wrapperMount.find(RecipeFull).length).toEqual(1);
    });

    it('renders Loading', ()  =>
    {
        expect(wrapperMount.find(Loading).length).toEqual(1);
    });

    it('renders Header', () =>
    {
        expect(wrapperMount.find(Header).length).toEqual(1);
    });

    it('renders Footer', () =>
    {
        expect(wrapperMount.find(Footer).length).toEqual(1);
    });

    it('renders HomeHeading', () =>
    {
        wrapperMount.setState({ results: mockRecipe, homeHead: mockHome });
        //wrapper.instance().forceUpdate()
        expect(wrapperMount.find(HomeHeading).length).toEqual(1);
    });

    //simple function render checks
    it('tests a function viewRecipeClick', () =>
    {
        expect(wrapper.instance().viewRecipeClick()).toMatchSnapshot();
    });

    it('tests a function viewHome', () =>
    {
        expect(wrapper.instance().viewHome()).toMatchSnapshot();
    });

    it('tests a function viewRecipeNavClick', () =>
    {
        expect(wrapper.instance().viewRecipeNavClick()).toMatchSnapshot();
    });

    it('tests a function changeRecipe', () =>
    {
        expect(wrapper.instance().changeRecipe()).toMatchSnapshot();
    });

    //check changeRecipe function when the passed value meets
    //its upper and lower limits
    it('test function changeRecipe when ID === -1', () =>
    {
        wrapper.setState({ results: mockRecipe, currentRecipe: 0 });
        wrapper.update();
        expect(wrapper.instance().changeRecipe(-1)).toEqual(mockRecipe.length - 1);
    });

    it('test function changeRecipe when ID === results.length', () =>
    {
        wrapper.setState({ results: mockRecipe, currentRecipe: mockRecipe.length - 1 });
        wrapper.update();
        expect(wrapper.instance().changeRecipe(mockRecipe.length)).toEqual(0);
    });

    //failsafe to close any asynchronous tasks
    afterAll(done => 
    {
        done()
    })
});

/* Recipe Component */
describe('Recipe', () => 
{
    let wrapper;    //shallow Recipe wrapper
    const setState = jest.fn();     //mock functions
    const useStateSpy = jest.spyOn(React, 'useState')       //state spy
    useStateSpy.mockImplementation((init) => [init, setState]);     //mimic state change

    //fill shallow component with mock data
    beforeEach(() => {
        wrapper = shallow(<Recipe
            ID={1}
            title={'title'}
            field_images={'field_images'}
            field_ingredients={'field_ingredients'}
            body={'body'}
            viewRecipeClick={setState}
        />);
    });

    //renders via snapshot
    test('Recipe snapshot renders', () => 
    {
        const component = renderer.create(<Recipe />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    //check that the button to view a recipe's details is clicked
    //and that it returns correct values
    it('view more button is clicked', () => 
    {
        wrapper.find('#full').props().onClick();
        expect(setState).toHaveBeenCalledWith(1);
    });
});

/* RecipeFull Component */
describe('RecipeFull', () => 
{
    //mock setup
    let wrapper;
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState')
    useStateSpy.mockImplementation((init) => [init, setState]);

    //populate component with mockdata
    beforeEach(() => {
        wrapper = shallow(<RecipeFull
            ID={1}
            title={'title'}
            field_images={'field_images'}
            field_ingredients={'field_ingredients'}
            body={'body'}
            changeRecipe={setState}
        />);
    });

    //clear mock data after each tests
    afterEach(() => {
        jest.clearAllMocks();
    });

    //test component renders via snapshot
    test('Recipe Full snapshot renders', () => 
    {
        const component = renderer.create(<RecipeFull />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    //check button clicks for previous and next recipe views
    //ensure the value in state is changed correctly
    // ID, title, field_images, field_ingredients, body, changeRecipe
    it('next button is clicked', () => 
    {
        wrapper.find('#next').props().onClick();
        expect(setState).toHaveBeenCalledWith(2);
    });

    it('previous button is clicked', () => 
    {
        wrapper.find('#previous').props().onClick();
        expect(setState).toHaveBeenCalledWith(0);
    });
    
});

/* HomeHeading Component */
/* Coverage Complete */
describe('HomeHeading', () => 
{
    test('HomeHeading snapshot renders', () => 
    {
        const component = renderer.create(<HomeHeading />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

/* Footer Component */
/* Coverage Complete */
describe('Footer', () => 
{
    test('Footer snapshot renders', () => 
    {
        const component = renderer.create(<Footer />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

/* Header Component */
describe('Header', () => 
{
    let navComponent;
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState')
    useStateSpy.mockImplementation((init) => [init, setState]);

    //fill component with mock data
    beforeEach(() => {
        navComponent = shallow(<Header
            viewHome={setState}
            viewRecipeNavClick={setState}
        />);
    });

    //check that it renders via snapshot
    test('Header snapshot renders', () => 
    {
        const component = renderer.create(<Header />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    //check for all nav.links
    it('should contain 3 nav.links', () => 
    {
        const items = navComponent.find(Nav.Link);
        expect(items).toHaveLength(3);
    });

    //check that clicking the home nav.link returns an empty call
    it('viewHome link is clicked', () => 
    {
        navComponent.find('#viewHome').props().onClick();
        expect(setState).toHaveBeenCalledWith();
    });

    //check that clicking the recipe list nav.link returns an empty call
    it('viewRecList link is clicked', () => 
    {
        navComponent.find('#viewRecList').props().onClick();
        expect(setState).toHaveBeenCalledWith();
    });

    //check that one navbar.brand renders
    it('should contain 1 navbar.brand', () => 
    {
        const items = navComponent.find(Navbar.Brand);
        expect(items).toHaveLength(1);
    });

    //check that clicking the home nav.link returns an empty call
    it('brand link is clicked', () => 
    {
        navComponent.find('#brand').props().onClick();
        expect(setState).toHaveBeenCalledWith();
    });
});

/* Loading Component */
/* Coverage Complete */
describe('Loading', () => 
{
    test('Loading snapshot renders', () => 
    {
        const component = renderer.create(<Loading />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
