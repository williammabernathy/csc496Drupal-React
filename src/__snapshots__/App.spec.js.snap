// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App snapshot renders 1`] = `
<div
  className="App"
>
  <div
    className="content"
  >
    <nav
      className="mainMenu navbar navbar-expand-lg navbar-light bg-light"
    >
      <a
        className="navBrand navbar-brand"
        href="#home"
        id="brand"
        onClick={[Function]}
      >
        CSC496
      </a>
      <div
        className="navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="navLink nav-link"
          data-rb-event-key="#home"
          href="#home"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          Home
        </a>
        <a
          className="navLink nav-link"
          data-rb-event-key="#recipe"
          href="#recipe"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          Recipes
        </a>
        <a
          className="navLink nav-link"
          data-rb-event-key="https://github.com/williammabernathy/csc496Drupal-React"
          href="https://github.com/williammabernathy/csc496Drupal-React"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          Github
        </a>
      </div>
    </nav>
    <div
      className="error"
    >
      Loading..
    </div>
    <div
      className="footer"
    >
      <span>
        CSC 496 Project
      </span>
       
      <br />
      <span>
        by William Abernathy
      </span>
    </div>
  </div>
</div>
`;

exports[`App tests a function changeRecipe 1`] = `undefined`;

exports[`App tests a function viewHome 1`] = `undefined`;

exports[`App tests a function viewRecipeClick 1`] = `undefined`;

exports[`App tests a function viewRecipeNavClick 1`] = `undefined`;

exports[`Footer Footer snapshot renders 1`] = `
<div
  className="footer"
>
  <span>
    CSC 496 Project
  </span>
   
  <br />
  <span>
    by William Abernathy
  </span>
</div>
`;

exports[`Header Header snapshot renders 1`] = `
<nav
  className="mainMenu navbar navbar-expand-lg navbar-light bg-light"
>
  <a
    className="navBrand navbar-brand"
    href="#home"
    id="brand"
    onClick={[Function]}
  >
    CSC496
  </a>
  <div
    className="navbar-nav"
    onKeyDown={[Function]}
  >
    <a
      className="navLink nav-link"
      data-rb-event-key="#home"
      href="#home"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
      Home
    </a>
    <a
      className="navLink nav-link"
      data-rb-event-key="#recipe"
      href="#recipe"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
      Recipes
    </a>
    <a
      className="navLink nav-link"
      data-rb-event-key="https://github.com/williammabernathy/csc496Drupal-React"
      href="https://github.com/williammabernathy/csc496Drupal-React"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
      Github
    </a>
  </div>
</nav>
`;

exports[`HomeHeading HomeHeading snapshot renders 1`] = `
<div>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": undefined,
      }
    }
  />
</div>
`;

exports[`Loading Loading snapshot renders 1`] = `
<div
  className="error"
>
  Loading..
</div>
`;

exports[`Recipe Recipe snapshot renders 1`] = `
<div
  className="recipeStyle"
>
  <div
    className="card card"
    style={
      Object {
        "border": "1px solid lightgrey",
        "width": "18rem",
      }
    }
  >
    <img
      alt="Food Image"
      className="card-img-top"
      src="http://gtest.dev.wwbtc.comundefined"
      width="25%"
    />
    <div
      className="card-body"
    >
      <div
        className="card-title"
      >
        <b />
      </div>
      <p
        className="card-text"
      >
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
        />
      </p>
      <center>
        <a
          aria-label={null}
          className="btn btn-secondary"
          href="#recipe/undefined"
          id="full"
          onClick={[Function]}
        >
          View Recipe
        </a>
      </center>
    </div>
  </div>
</div>
`;

exports[`RecipeFull Recipe Full snapshot renders 1`] = `
<div
  className="cardFull"
>
  <div
    className="cardFullImage"
  >
    <img
      alt="Food"
      src="http://gtest.dev.wwbtc.comundefined"
    />
  </div>
  <div>
    <hr />
    <div>
      <b />
    </div>
    <br />
    <div />
    <br />
    <div>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": undefined,
          }
        }
      />
    </div>
    <hr />
    <a
      aria-label={null}
      className="previous btn btn-secondary"
      href="#recipe/undefined"
      id="previous"
      onClick={[Function]}
    >
      Previous
    </a>
    <a
      aria-label={null}
      className="next btn btn-secondary"
      href="#recipe/undefined"
      id="next"
      onClick={[Function]}
    >
      Next
    </a>
  </div>
</div>
`;
